Shell/AST.h

```diff
@@ -0,0 +1,730 @@
+};
+
+struct Rewiring {
+    int source_fd { 0 };
+    int dest_fd { 0 };
+    enum class Close {
+        None,
+        Source,
+        Destination,
+    } must_be_closed { Close::None };
+};
+
+struct Command {
+    Vector<String> argv;
+    Vector<Redirection> redirections;
+    Vector<Rewiring> rewirings;
+    bool should_be_run_in_background { false };
+    bool is_pipe_source { false };
+};
+
+struct HitTestResult {
+    RefPtr<Node> matching_node;
+    RefPtr<Node> closest_node_with_semantic_meaning; // This is used if matching_node is a bareword
+};
+
+class Value : public RefCounted<Value> {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) = 0;
+    virtual Vector<Command> resolve_as_commands(TheExecutionInputType);
+    virtual ~Value();
+    virtual bool is_command() const { return false; }
+    virtual bool is_glob() const { return false; }
+    virtual bool is_job() const { return false; }
+    virtual bool is_list() const { return false; }
+    virtual bool is_string() const { return false; }
+};
+
+class CommandValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual Vector<Command> resolve_as_commands(TheExecutionInputType) override;
+    virtual ~CommandValue();
+    virtual bool is_command() const override { return true; }
+    CommandValue(Command command)
+        : m_command(move(command))
+    {
+    }
+
+    CommandValue(Vector<String> argv)
+        : m_command({ move(argv), {}, {}, false })
+    {
+    }
+
+private:
+    Command m_command;
+};
+
+class CommandSequenceValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual Vector<Command> resolve_as_commands(TheExecutionInputType) override;
+    virtual ~CommandSequenceValue();
+    virtual bool is_command() const override { return true; }
+    CommandSequenceValue(Vector<Command> commands)
+        : m_contained_values(move(commands))
+    {
+    }
+
+private:
+    Vector<Command> m_contained_values;
+};
+
+class JobValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override { ASSERT_NOT_REACHED(); }
+    virtual Vector<Command> resolve_as_commands(TheExecutionInputType) override { ASSERT_NOT_REACHED(); }
+    virtual ~JobValue();
+    virtual bool is_job() const override { return true; }
+    JobValue(RefPtr<Job> job)
+        : m_job(move(job))
+    {
+    }
+
+    const RefPtr<Job> job() const { return m_job; }
+
+private:
+    RefPtr<Job> m_job;
+};
+
+class ListValue : public Value {
+public:
+    virtual Vector<String> resolve_as_list(TheExecutionInputType) override;
+    virtual ~ListValue();
+    virtual bool is_list() const override { return true; }
+    ListValue(Vector<RefPtr<Value>> values)
+        : m_contained_values(move(values))
+    {
+    }
+
+private:
+    Vector<RefPtr<Value>> m_contained_values;
+};
+
+class StringListValue : public Value {
```

Comment:
Wait how is "string list" different from just "list"?

Reply:
`StringList` stores `String`s, `List` stores `Value`s.

Reply:
I get as much :slightly_smiling_face:
but why'd we have a separate type for a list of strings? is this for the implementation efficiency, or does it impact shell semantics?

Reply:
Nah, it's just for efficiency.
also allows us to not add `resolve_as_string`, one less virtual thing to implement in a billion places.

Reply:
Then perhaps add a comment?

`// A list that only contains strings, specialized for efficiency.`